workspaces/orchestrator/plugins/orchestrator-form-widgets/src/FormWidgetsApi.tsx

```diff
@@ -0,0 +1,162 @@
+  pathPrefix: string,
+) => void = (uiSchema, formData, callback, pathPrefix) => {
+  let dottedPathPrefix = pathPrefix;
+  if (pathPrefix) {
+    dottedPathPrefix += '.';
+  }
+
+  for (const key of Object.keys(uiSchema)) {
+    // tune this condition to match just those properties which are relevant for the orchestrator-form-widget
+    if (typeof uiSchema[key] === 'object') {
+      if (uiSchema[key]?.['ui:widget']) {
+        callback(`${dottedPathPrefix}${key}`, uiSchema[key], formData?.[key]);
+      } else {
+        walkThrough(
+          uiSchema[key],
+          formData?.[key] as JsonObject,
+          callback,
+          `${dottedPathPrefix}${key}`,
+        );
+      }
+    }
+  }
+};
+
+const safeSet: (errors: JsonObject, path: string, value: JsonValue) => void = (
+  errors,
+  path,
+  value,
+) => {
+  const steps = path.split('.', 2);
+  if (steps.length === 1) {
+    errors[steps[0]] = value;
+  } else {
+    const safeObject = (errors[steps[0]] ?? {}) as JsonObject;
+    errors[steps[0]] = safeObject;
+    safeSet(safeObject, steps[1], value);
+  }
+};
+export class FormWidgetsApi implements OrchestratorFormApi {
+  // private readonly configApi: ConfigApi;
+  // private readonly fetchApi: FetchApi;
+
+  // public constructor(options: { configApi: ConfigApi; fetchApi: FetchApi }) {
+  //   this.configApi = options.configApi;
+  //   this.fetchApi = options.fetchApi;
+  // }
+
+  getFormDecorator: OrchestratorFormApi['getFormDecorator'] = () => {
+    return (FormComponent: React.ComponentType<FormDecoratorProps>) => {
+      return () => {
+        const { formData, setFormData, uiSchema } =
+          useWrapperFormPropsContext();
+
+        const widgets = { SchemaUpdater, ActiveTextInput };
+
+        const onChange = useCallback(
+          (data: JsonObject | undefined) => {
+            if (data) {
+              setFormData(data);
+            }
+          },
+          [setFormData],
+        );
+
+        const getExtraErrors: (
+          currentFormData: JsonObject,
+        ) => Promise<ErrorSchema<JsonObject>> = async (
+          currentFormData: JsonObject,
+        ) => {
+          // Asynchronous validation on wizard step transition or submit
+          return sleep(1000 /* The sleep mimics async fetch, remove it */).then(
+            () => {
+              const errors: ErrorSchema<JsonObject> = {};
+              walkThrough(
+                uiSchema,
+                currentFormData,
+                (path, uiSchemaProperty, formDataValue) => {
+                  if (uiSchemaProperty?.['ui:widget'] === 'ActiveTextInput') {
+                    if (formDataValue) {
+                      safeSet(errors, path, {
+                        __errors: [
+                          'This is a dummy error from async validator',
+                          `Yet another issue with this field of value: ${formDataValue}`,
+                          'Remove the value if you want this silly validator to pass',
```

Comment:
are these needed?

Reply:
Right, the walk-thought video has be taken, we can remove it.